### PR TITLE
A few fixes

### PR DESCRIPTION
**Pull request by vsch**
- fix: lexer allowed key names to span more than one line, embedding CRLF in the key name. 

    As a challenge, I first tried to change the code to work properly with embedded LF in key names but found it impossible to get everything working right. Gave up and changed the grammar to not allow it. 

- fix: keys with spaces were not handled properly by code completion, structure view, and create new property quick fix. 

    Code completion and structure view would have the escaped spaces instead of plain spaces. Quick fix needed to escape any spaces in keys otherwise only the first word of the key name would be inserted as the new property key.

- fix: folding of keys whose values had embedded \n now display them as \n so they don't look like multiline strings. Makes it visually consistent with embedded \n in strings.

- fix: add tests for null key value, to handle keys with no values since the grammar allows it. 

- fix: filtering of child crlf nodes so that max blank lines in code style is respected. Previously it would be 1+ the setting.

- add commented out code in quick fix to do the same thing as code using PSI but by manipulating plain text of the document, just as an example of manipulating document text.

### CODE REVIEW

File: src/com/simpleplugin/Simple.flex [hunk 1]
```diff
@@ -21,7 +21,7 @@ FIRST_VALUE_CHARACTER=[^ \n\r\f\\] | "\\"{CRLF} | "\\".
 VALUE_CHARACTER=[^\n\r\f\\] | "\\"{CRLF} | "\\".
 END_OF_LINE_COMMENT=("#"|"!")[^\r\n]*
 SEPARATOR=[:=]
-KEY_CHARACTER=[^:=\ \n\r\t\f\\] | "\\"{CRLF} | "\\".
+KEY_CHARACTER=[^:=\ \n\r\t\f\\] | "\\ "
```
Comment (vsch) [inline src/com/simpleplugin/Simple.flex]: disallow embedding \n in key names

File: src/com/simpleplugin/SimpleFoldingBuilder.java [hunk 2]
```diff
@@ -38,7 +45,10 @@
                         @Nullable
                         @Override
                         public String getPlaceholderText() {
-                            return properties.get(0).getValue();
+                            // IMPORTANT: keys can come with no values, so a test for null is needed
+                            // IMPORTANT: Convert embedded \n to backslash n, so that the string will look like it has LF embedded in it and embedded " to escaped "
+                            String valueOf = properties.get(0).getValue();
+                            return valueOf == null ? "" : valueOf.replaceAll("\n","\\n").replaceAll("\"","\\\\\"");
```
Comment (vsch) [inline src/com/simpleplugin/SimpleFoldingBuilder.java]: escape embedded \n and "

**Comment by cheptsov** [comment]
Could you please add tests covering your changes?

**Comment by vsch** [comment]
Sure. I'll do that. 

Best regards,

Vladimir
(514) 244-4172

Sent from my iPhone

On Aug 17, 2015, at 3:11 AM, Andrey Cheptsov <notifications@github.com> wrote:

Could you please add tests covering your changes?

**Comment by vsch** [comment]
###### Tests added. More fixes added.

- fix check for String type for literal elements before casting to String

- add test for key names that have spaces and values that have more than one line, to make the tests more interesting

- add format KEEP_BLANK_LINES_IN_CODE = 1, this test fails. Formatting in test and in settings code style preview not applying the format. Format works when invoked in a file.

I noticed this when I was working with formatting. The code style pane does not reflect the changes in keep blank lines in code. It does do something because it flashes the changes and sometimes inserts/deletes lines but never correctly, except by accident when the setting equals the number of blank lines in the sample.

However, **when applied to a document the formatting works correctly.** 

**Any advice on what I need to do to make formatting work correctly for the test, and I am assuming the same will fix the code style preview pane, will be appreciated.**

For this plug-in I am working with IntelliJ IDEA CE 141.2311.

image
 
During tests I got some extra errors that local history is broken. 

```
/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/bin/java -ea -Xbootclasspath/p:../out/classes/production/boot -XX:+HeapDumpOnOutOfMemoryError -Xmx512m -XX:MaxPermSize=320m -Didea.system.path=../test-system -Didea.home.path=. -Didea.config.path=../test-config -Didea.test.group=ALL_EXCLUDE_DEFINED "-Xbootclasspath/a:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/boot.jar" -Didea.load.plugins.id=com.simpleplugin.unique.plugin.id -Didea.plugins.path=/Users/vlad/Library/Caches/IdeaIC14/plugins-sandbox/plugins -Didea.launcher.port=7532 "-Didea.launcher.bin.path=/Applications/IntelliJ IDEA 14 CE.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath "/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/tools.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/plugins/junit/lib/junit-rt.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/annotations.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/asm-all.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/asm-commons.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/asm.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/asm4-all.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/automaton.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/boot.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/bootstrap.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/cglib-2.2.2.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/cli-parser-1.1.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/commons-codec-1.8.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/commons-httpclient-3.1-patched.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/commons-logging-1.1.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/commons-net-3.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/ecj-4.4.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/extensions.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/fluent-hc-4.3.6.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/forms_rt.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/freemarker.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/groovy-all-2.3.9.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/gson-2.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/guava-17.0.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/hamcrest-core-1.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/hamcrest-library-1.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/httpclient-4.3.6.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/httpcore-4.3.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/httpmime-4.3.6.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/icons.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/idea.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/isorelax.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/javac2.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jaxen-1.1.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jayatana-1.2.4.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jcip-annotations.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jdkAnnotations.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jdom.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jgoodies-common-1.2.1.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jgoodies-forms.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jgoodies-looks-2.4.2.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jh.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jing.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jna-utils.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jna.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jps-launcher.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jps-model.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jps-server.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch-0.1.51.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.connector-factory.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.core.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.pageant.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.sshagent.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.usocket-jna.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsch.agentproxy.usocket-nc.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsr166e.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jsr173_1.0_api.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/junit-4.11.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/jzlib-1.1.1.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/log4j.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/markdown4j-2.2.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/markdownj-core-0.4.2-SNAPSHOT.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/microba.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/miglayout-swing.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/nanoxml-2.2.3.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/nekohtml-1.9.14.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/netty-all-4.1.0.Beta4.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/openapi.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/optimizedFileManager.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/oromatcher.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/picocontainer.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/protobuf-2.5.0.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/proxy-vole_20131209.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/pty4j-0.4.15.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/purejavacomm.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/resolver.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/resources.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/resources_en.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/rhino-js-1_7R4.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/rngom-20051226-patched.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/sanselan-0.98-snapshot.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/serviceMessages.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/slf4j-api-1.7.10.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/slf4j-log4j12-1.7.10.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/snappy-in-java-0.3.1.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/swingx-core-1.6.2.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/trang-core.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/trove4j.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/trove4j_src.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/util.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/velocity.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/winp-1.21-patched.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/xbean.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/xerces.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/xmlrpc-2.0.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/xpp3-1.1.4-min.jar:/Applications/IntelliJ IDEA 14 CE.app/Contents/lib/xstream-1.4.3.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Classes/charsets.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Classes/classes.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Classes/jsse.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Classes/ui.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/ext/apple_provider.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdl1.6.0.jdk/Contents/Home/lib/ext/sunpkcs11.jar:/Users/vlad/src/SimplePlugin/out/test/SimplePlugin:/Users/vlad/src/SimplePlugin/out/production/SimplePlugin" com.intellij.rt.execution.application.AppMain com.intellij.rt.execution.junit.JUnitStarter -ideVersion5 @/private/var/folders/bs/4ktxsyn54pj5mc0l8b43h8vc0000gp/T/idea_junit0.tmp @w@/private/var/folders/bs/4ktxsyn54pj5mc0l8b43h8vc0000gp/T/idea_working_dirs_junit0.tmp -socket52271
[13:30:53,312]   INFO - api.vfs.impl.local.FileWatcher - Native file watcher is disabled 

com.intellij.rt.execution.junit.FileComparisonFailure: DefaultTestData.simple     <Click to see difference>



	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.checkResult(CodeInsightTestFixtureImpl.java:1868)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.checkResultByFile(CodeInsightTestFixtureImpl.java:1837)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.access$1000(CodeInsightTestFixtureImpl.java:128)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl$19.run(CodeInsightTestFixtureImpl.java:1174)
	at com.intellij.util.ui.UIUtil.invokeAndWaitIfNeeded(UIUtil.java:2369)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.checkResultByFile(CodeInsightTestFixtureImpl.java:1170)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.checkResultByFile(CodeInsightTestFixtureImpl.java:1164)
	at com.simpleplugin.SimpleCodeInsightTest.testFormatter(SimpleCodeInsightTest.java:48)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at com.intellij.testFramework.UsefulTestCase.access$001(UsefulTestCase.java:82)
	at com.intellij.testFramework.UsefulTestCase$2.run(UsefulTestCase.java:302)
	at com.intellij.util.ui.UIUtil.invokeAndWaitIfNeeded(UIUtil.java:2369)
	at com.intellij.testFramework.UsefulTestCase.invokeTestRunnable(UsefulTestCase.java:334)
	at com.intellij.testFramework.UsefulTestCase.runTest(UsefulTestCase.java:318)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:346)
	at com.intellij.testFramework.UsefulTestCase$4.run(UsefulTestCase.java:428)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:199)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:715)
	at java.awt.EventQueue.access$400(EventQueue.java:82)
	at java.awt.EventQueue$2.run(EventQueue.java:676)
	at java.awt.EventQueue$2.run(EventQueue.java:674)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.AccessControlContext$1.doIntersectionPrivilege(AccessControlContext.java:86)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:685)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:362)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:296)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:211)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:196)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:188)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:122)

ERROR: Local history is broken(version:5,current timestamp:Aug 17, 2015 1:31:01 PM,storage timestamp:Aug 17, 2015 1:30:53 PM,vfs timestamp:Aug 17, 2015 1:30:53 PM)
null
java.io.IOException: Recursive records found
	at com.intellij.history.core.ChangeListStorageImpl.doReadPrevSafely(ChangeListStorageImpl.java:280)
	at com.intellij.history.core.ChangeListStorageImpl.findFirstObsoleteBlock(ChangeListStorageImpl.java:271)
	at com.intellij.history.core.ChangeListStorageImpl.purge(ChangeListStorageImpl.java:237)
	at com.intellij.history.core.ChangeList.purgeObsolete(ChangeList.java:175)
	at com.intellij.history.integration.LocalHistoryImpl.disposeComponent(LocalHistoryImpl.java:141)
	at com.intellij.history.integration.LocalHistoryImpl$1.run(LocalHistoryImpl.java:64)
	at com.intellij.openapi.util.ShutDownTracker.run(ShutDownTracker.java:62)
	at java.lang.Thread.run(Thread.java:695)
ERROR: Local history is broken(version:5,current timestamp:Aug 17, 2015 1:31:01 PM,storage timestamp:Aug 17, 2015 1:30:53 PM,vfs timestamp:Aug 17, 2015 1:30:53 PM)
null
java.lang.AssertionError: Local history is broken(version:5,current timestamp:Aug 17, 2015 1:31:01 PM,storage timestamp:Aug 17, 2015 1:30:53 PM,vfs timestamp:Aug 17, 2015 1:30:53 PM)
null
	at com.intellij.testFramework.LoggedErrorProcessor.processError(LoggedErrorProcessor.java:56)
	at com.intellij.testFramework.TestLogger.error(TestLogger.java:67)
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:130)
	at com.intellij.history.core.ChangeListStorageImpl.handleError(ChangeListStorageImpl.java:110)
	at com.intellij.history.core.ChangeListStorageImpl.purge(ChangeListStorageImpl.java:250)
	at com.intellij.history.core.ChangeList.purgeObsolete(ChangeList.java:175)
	at com.intellij.history.integration.LocalHistoryImpl.disposeComponent(LocalHistoryImpl.java:141)
	at com.intellij.history.integration.LocalHistoryImpl$1.run(LocalHistoryImpl.java:64)
	at com.intellij.openapi.util.ShutDownTracker.run(ShutDownTracker.java:62)
	at java.lang.Thread.run(Thread.java:695)
Exception in thread "Shutdown tracker" java.lang.AssertionError: Local history is broken(version:5,current timestamp:Aug 17, 2015 1:31:01 PM,storage timestamp:Aug 17, 2015 1:30:53 PM,vfs timestamp:Aug 17, 2015 1:30:53 PM)
null
	at com.intellij.testFramework.LoggedErrorProcessor.processError(LoggedErrorProcessor.java:56)
	at com.intellij.testFramework.TestLogger.error(TestLogger.java:67)
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:134)
	at com.intellij.openapi.util.ShutDownTracker.run(ShutDownTracker.java:65)
	at java.lang.Thread.run(Thread.java:695)

Process finished with exit code 255
```

File: testData/FindUsagesTestData.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 public class Test {
     public static void main(String[] args) {
-        System.out.println("simple:website");
+        System.out.println("simple:key with spaces");
```
Comment (vsch) [inline testData/FindUsagesTestData.java]: take a more interesting key name for the test

File: testData/FoldingTestData.java [hunk 4]
```diff
@@ -2,4 +2,10 @@
     public static void main(String[] args)<fold text=' { '> {
         </fold>System.out.println("<fold text='http://en.wikipedia.org/'>simple:website</fold>");<fold text=' }'>
     }</fold>
+    public static void main1(String[] args)<fold text=' { '> {
+        </fold>System.out.println("<fold text='This is the value that could be looked up with the key \"key with spaces\".'>simple:key with spaces</fold>");<fold text=' }'>
+    }</fold>
+    public static void main2(String[] args)<fold text=' { '> {
+        </fold>System.out.println("<fold text='Welcome to \n          Wikipedia!'>simple:message</fold>");<fold text=' }'>
+    }</fold>
```
Comment (vsch) [inline testData/FoldingTestData.java]: add more complex key names and value for the folding test

File: testData/FormatterTestData.simple [hunk 5]
```diff
@@ -2,6 +2,8 @@
 ! The exclamation mark can also mark text as comments.
 website=http://en.wikipedia.org/
 
+
+
```
Comment (vsch) [inline testData/FormatterTestData.simple]: add extra blank lines that should be removed by formatting test

File: testData/FormatterTestData.simple [hunk 6]
```diff
@@ -10,4 +12,4 @@ message  =  Welcome to \
 # Add spaces to the key
 key\ with\ spaces = This is the value that could be looked up with the key "key with spaces".
 # Unicode
-tab :\u0009
\ No newline at end of file
+tab :\u0009
```
Comment (vsch) [inline testData/FormatterTestData.simple]: add one blank line that should be left after formatting

File: testData/ParsingTestData.simple [hunk 7]
```diff
@@ -10,4 +10,8 @@ message = Welcome to \
 # Add spaces to the key
 key\ with\ spaces = This is the value that could be looked up with the key "key with spaces".
 # Unicode
-tab : \u0009
\ No newline at end of file
+tab : \u0009
+# test for illegal key attempt
+key\
+with\
+endofline = test
```
Comment (vsch) [inline testData/ParsingTestData.simple]: attempt to embed \n in key name

File: META-INF/plugin.xml [hunk 8]
```diff
@@ -40,7 +40,7 @@
     <!-- Add your extensions here -->
       <fileTypeFactory implementation="com.simpleplugin.SimpleFileTypeFactory"/>
       <lang.parserDefinition language="Simple" implementationClass="com.simpleplugin.SimpleParserDefinition"/>
-      <lang.syntaxHighlighterFactory key="Simple" implementationClass="com.simpleplugin.SimpleSyntaxHighlighterFactory"/>
+      <lang.syntaxHighlighterFactory key="Simple" language="Simple" implementationClass="com.simpleplugin.SimpleSyntaxHighlighterFactory"/>
```
Comment (vsch) [inline META-INF/plugin.xml]: Version 14.1 requires key=..., Version 14.2 only works with language=.....
Setting both allows the plugin to work in different versions.

**Comment by vsch** [comment]
Added a todo indexer to show a sample of participating in todo item population in a plugin.

**Comment by cheptsov** [comment]
Indeed, the formatter has been working incorrectly. I've fixed this problem by changing in the lexer: https://github.com/JetBrains/SimplePlugin/commit/0673274a398a2f2d895f575b1e369550f874dc31

Could you please rebase? Sorry but it was easier to fix it in my version.

**Comment by vsch** [comment]
No need to apologize. I prefer to see your fixes to get a better understanding of how it should be done. I am still in the neophyte stage of IDEA plugin development. I will rebase.

File: META-INF/plugin.xml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <idea-plugin version="2">
-  <id>com.yourcompany.unique.plugin.id</id>
-  <name>Plugin display name here</name>
+  <id>com.simpleplugin.unique.plugin.id</id>
+  <name>Simple</name>
```
Comment (vsch) [inline META-INF/plugin.xml]: Change to reflect the plugin name.

**Comment by vsch** [comment]
Can be merged. I changed the location of my community edition so that the working dir is the same as yours relative to the plugin and that fixed the history complaint while running tests too.
